nntrainer/layers/lstm.cpp

```diff
@@ -218,11 +223,118 @@ void LSTMLayer::copy(std::shared_ptr<Layer> l) {
+  Tensor &input_ = net_input[0]->getVariableRef();
+  Tensor &m_cell_ = mem_cell->getVariableRef();
+  Tensor &dm_cell_ = mem_cell->getGradientRef();
+
+  Tensor dh_nx = Tensor({derivative_.width()});
+  Tensor dc_nx = Tensor({derivative_.width()});
+
+  for (unsigned int b = 0; b < input_dim[0].batch(); ++b) {
+    Tensor deriv_t = derivative_.getBatchSlice(b, 1);
+    Tensor derivc_t = dm_cell_.getBatchSlice(b, 1);
+    Tensor xs_t = input_.getBatchSlice(b, 1);
+    Tensor hs_t = hidden_.getBatchSlice(b, 1);
+    Tensor cs_t = m_cell_.getBatchSlice(b, 1);
+
+    dh_nx.setZero();
+    dc_nx.setZero();
+
+    Tensor dh;
+    Tensor xs;
+    Tensor hs_prev;
+    Tensor cs_prev;
+    Tensor hs;
+    Tensor cs;
+    Tensor dc;
+    Tensor dfgio_ = fgio->getGradientRef().getBatchSlice(b, 1);
+    Tensor fgio_ = fgio->getVariableRef().getBatchSlice(b, 1);
+
+    for (unsigned int t = deriv_t.height(); t-- > 0;) {
+      dh = deriv_t.getSharedDataTensor({deriv_t.width()}, t * deriv_t.width());
+      dc =
+        derivc_t.getSharedDataTensor({derivc_t.width()}, t * derivc_t.width());
+      xs = xs_t.getSharedDataTensor({xs_t.width()}, t * xs_t.width());
+      hs = hs_t.getSharedDataTensor({hs_t.width()}, t * hs_t.width());
+      cs = cs_t.getSharedDataTensor({cs_t.width()}, t * cs_t.width());
+
+      Tensor dfgio_t =
+        dfgio_.getSharedDataTensor({unit * NUM_GATE}, unit * t * NUM_GATE);
+      Tensor fgio_t =
+        fgio_.getSharedDataTensor({unit * NUM_GATE}, unit * t * NUM_GATE);
+
+      if (t == 0) {
+        hs_prev = Tensor({hs_t.width()});
+        hs_prev.setZero();
+        cs_prev = Tensor({cs_t.width()});
+        cs_prev.setZero();
+      } else {
+        hs_prev =
+          hs_t.getSharedDataTensor({hs_t.width()}, (t - 1) * hs_t.width());
+        cs_prev =
+          cs_t.getSharedDataTensor({cs_t.width()}, (t - 1) * cs_t.width());
+      }
+
+      if (t < deriv_t.height() - 1) {
+        dh.add_i(dh_nx);
+      }
+
+      Tensor dhf = dfgio_t.getSharedDataTensor({unit}, 0);
+      Tensor dhg = dfgio_t.getSharedDataTensor({unit}, unit);
+      Tensor dhi = dfgio_t.getSharedDataTensor({unit}, unit * 2);
+      Tensor dho = dfgio_t.getSharedDataTensor({unit}, unit * 3);
+
+      Tensor hf = fgio_t.getSharedDataTensor({unit}, 0);
+      Tensor hg = fgio_t.getSharedDataTensor({unit}, unit);
+      Tensor hi = fgio_t.getSharedDataTensor({unit}, unit * 2);
+      Tensor ho = fgio_t.getSharedDataTensor({unit}, unit * 3);
+
+      recurrent_acti_func.run_fn(cs, dho);
+      dho.multiply_i(dh);
+      recurrent_acti_func.run_prime_fn(cs, dc, ho);
+      dc.multiply_i(dh);
+      dc.add_i(dc_nx);
+
+      dc.multiply(cs_prev, dhf);
+      dc.multiply(hg, dhi);
+      dc.multiply(hi, dhg);
+      dc.multiply(hf, dc_nx);
+
+      acti_func.run_prime_fn(ho, dho, dho);
+      acti_func.run_prime_fn(hf, dhf, dhf);
+      acti_func.run_prime_fn(hi, dhi, dhi);
+      recurrent_acti_func.run_prime_fn(hg, dhg, dhg);
```

Comment:
Could you check if this is okay?

`acti_func.run_prime_fn(a, b, b)`
```cpp
prim_fn(x, y, z) {
1: run_prim_fn(x, y)
2: return y * z; 
}
```
y is changed at line 1, in this case z is same as y so the data is changed.

Reply:
From the offline discussion, I think this is beyond the scope of this PR. 

Since z is passed as `const` I think we shouldn't allow changing z. Let's deal this problem later :)

Reply:
Checked. It's Ok.